### PR TITLE
Minimize search source of shard level search requests

### PR DESCRIPTION
**Pull request by dnhatn**
This change reduces the memory usage and serialization cost of shard-level search requests by discarding the point in time and collapse from search sources of these shard-level requests.

I will work on another PR to remove SearchContextId from search responses.

Relates #80187

### CODE REVIEW

**Comment by dnhatn** [comment]
run elasticsearch-ci/part-2

File: server/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java [hunk 1]
```diff
@@ -1809,4 +1809,18 @@ public String toString(Params params) {
...
 
+    /**
+     * Returns a minimal source that discards information that aren't required by shard-level search requests.
+     * The purpose is to minimize the memory usage and serialization cost of shard-level search requests.
+     */
+    public SearchSourceBuilder minimalSourceForShardRequests() {
+        if (pointInTimeBuilder != null || collapse != null) {
+            final SearchSourceBuilder source = shallowCopy();
+            source.pointInTimeBuilder(null);
+            source.collapse(null);
```
Comment (jimczi) [inline server/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java:1820]: The collapse builder is used to build the search context so we should not remove it ?

File: server/src/main/java/org/elasticsearch/search/internal/ShardSearchRequest.java [hunk 2]
```diff
@@ -392,6 +392,9 @@ public void setAliasFilter(AliasFilter aliasFilter) {
     }
 
     public void source(SearchSourceBuilder source) {
+        if (source != null) {
+            source = source.minimalSourceForShardRequests();
```
Comment (jimczi) [inline server/src/main/java/org/elasticsearch/search/internal/ShardSearchRequest.java:396]: Should we call this on the constructor too ?

**Review by jimczi — COMMENTED** [review]
I left some comments.

File: server/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java (hunk 1 above)
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java]: Good catch. Fixed in https://github.com/elastic/elasticsearch/pull/80634/commits/d99725efd5f247c3ffe1e6b5c8f9e679ef43129e

File: server/src/main/java/org/elasticsearch/search/internal/ShardSearchRequest.java (hunk 2 above)
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/search/internal/ShardSearchRequest.java:396]: +1

**Comment by dnhatn** [comment]
@jimczi I've addressed your comments. Can you take another look? Thank you!

File: server/src/main/java/org/elasticsearch/search/internal/ShardSearchRequest.java [hunk 3]
```diff
@@ -236,7 +236,7 @@ public ShardSearchRequest(StreamInput in) throws IOException {
         shardRequestIndex = in.getVersion().onOrAfter(Version.V_7_11_0) ? in.readVInt() : -1;
         numberOfShards = in.readVInt();
         scroll = in.readOptionalWriteable(Scroll::new);
-        source = in.readOptionalWriteable(SearchSourceBuilder::new);
+        source(in.readOptionalWriteable(SearchSourceBuilder::new));
```
Comment (jimczi) [inline server/src/main/java/org/elasticsearch/search/internal/ShardSearchRequest.java:239]: Sorry I wasn't clear, I was referring to the other constructors above that takes the source as an argument. We should always create the request with the minimal source.

**Comment by dnhatn** [comment]
@elasticmachine update branch

**Review by jimczi — APPROVED** [review]
LGTM

**Comment by dnhatn** [comment]
Thanks Jim!

**Pull request merged**